mesonbuild/dependencies/cuda.py

```diff
@@ -0,0 +1,253 @@
+# Copyright 2013-2019 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import glob
+import re
+import os
+
+from .. import mlog
+from .. import mesonlib
+from ..environment import detect_cpu_family
+
+from .base import (DependencyException, ExternalDependency, HasNativeKwarg)
+
+
+class CudaDependency(ExternalDependency):
+
+    supported_languages = ['cuda', 'cpp', 'c'] # see also _default_language
+
+    def __init__(self, environment, kwargs):
+        HasNativeKwarg.__init__(self, kwargs) # initialize self.for_machine
+        compilers = environment.coredata.compilers[self.for_machine]
+        language = self._detect_language(compilers)
+        if language not in self.supported_languages:
+            raise DependencyException('Language \'{}\' is not supported by the CUDA Toolkit. Supported languages are {}.'.format(language, self.supported_languages))
+
+        super().__init__('cuda', environment, language, kwargs)
+        self.requested_modules = self.get_requested(kwargs)
+        if 'cudart' not in self.requested_modules:
+            self.requested_modules = ['cudart'] + self.requested_modules
```

Comment:
In https://github.com/elFarto/nvidia-vaapi-driver/issues/3#issuecomment-1005488544 there is apparently an objection to including libcudart in a use case where only libcuda is desired.

Since I don't know anything about the topic... Why is this necessary given you can already add it to the requested_modules via a kwarg?

Reply:
@eli-schwartz IIRC, I wanted the mainstream CUDA use case — which includes linking against the runtime (`cudart`) — to work out-of-the-box w/ as little boilerplate code on the user side as possible. That said, it looks like the way I did it does not allow one to override the default behavior, which is for sure a design bug. Please feel free to fix/revise.

Reply:
@elFarto thoughts?

Reply:
I don't recommend this. NVCC has the flag [`--cudart {none,shared,static}`](https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#file-and-path-specifications-cudart) (default: `static`) for just this purpose, and the reason why `none` is allowed is that an application that goes to the trouble of shipping or embedding CUBINs and using only the CUDA Driver API can avoid any interaction with the CUDA Runtime API and rely only on the driver to call kernels. That makes it CUDA Toolkit-independent, which is an important positive. Code can `static`-ally link `libcudart` to get most of these advantages without most of the pain, but that makes it unsuitable for linking with code compiled by other CUDA Toolkits.

Reply:
I agree mostly with the way @agurtovoy designed it, as it's the most common use case. However, hard coding the cudart module in there is not ideal. Maybe we could add something like (I don't know Python too well):
```
if 'cudart' not in self.requested_modules and self.requested_modules != ['cuda']:
```
This should keep the default behaviour, and still allow overriding when you only want the driver, not the runtime. I think that you'd either want just the driver, or you'd want the runtime + maybe some other modules.

Reply:
If I was designing this dependency today, I would definitely say "adding the module isn't really boilerplate IMO, so people should specify what they want and be allowed to not specify what they don't want".

The question is what to do about existing users... I'll defer to people who have a stake in the matter (and understand the stakes involved) to figure out the best way to detect the right thing to do. :)